### DISCUSSION TITLE
help with HIC ID issue

### DISCUSSION THREAD

**Discussion by andreipavlevich** [ARMmbed/DAPLink#956] [Q&A] [answered]
Hello,
I have a weird situation and looking for help or advice. I have 2 micro:bits v 1.5 running 0254 FW. I decided to flash 1 with the latest firmware but it stuck in maintenance mode. I tried many versions of FWs but nothing worked out. When I tried to find out what's the problem I found that both 2 micro:bits have HIC ID 97969908 instead of 97969901. The micro:bit with original FW that works fine has UID started with 99013601. The micro:bit that stuck in maintenance mode has UID started with 00000000. From the source code of DAPLink I found that HIC ID 97969908 is related to STM32F103 controller. I decided to flash micro:bit that stuck in maintenance mode with FW for STM32F103 and it flashed successfully.
I have no idea why my micro:bits have 'wrong' HIC ID but it looks like I'm not able to flash KL26Z FW because micro:bit doesn't pass HIC ID check. Also it looks like during the flashing with STM32 FW it changed first 8 digits of UID. 
If the problem is in ID check, can someone help me to build FW for micro:bit 1.5 with HID IC 97969908? Unfortunately I'm not a developer. Thanks in advance!

PS if one google 'micro:bit stuck in maintenance mode' you will see that users also have micro:bit with HIC ID 97969908.

**Comment by microbit-carlos** [comment]
If you send an email to help@microbit.org (or open a ticket via https://support.microbit.org) the micro:bit support team should be able to help.

**Reply by andreipavlevich** [reply]
> If you send an email to [help@microbit.org](mailto:help@microbit.org) (or open a ticket via https://support.microbit.org) the micro:bit support team should be able to help.

Hi Carlos, 
they already said I purchased counterfeit products (some mismatches in logo positioning). But this also indicates that it is possible to flash micro:bit firmware to device with wrong HIC ID. I believe I'm not alone who was cheated by unfair seller and purchased a counterfeit. I would be grateful for any kind of advice or help please.

**Comment by andreipavlevich** [comment]
As I wrote previously I have 2 counterfeit micro:bits v1.5. Both have M26M7V and N51822 chips.
![IMG_2799](https://user-images.githubusercontent.com/73068640/164972824-cef594c9-3ed2-4c70-a0b6-5d507b146df5.jpeg)

One of them is running well on 0254 FW that came with it. It appears as MICROBIT drive and works well with Makecode and with other original micro:bits v.2 I also have. Here is details.txt:

# DAPLink Firmware - see https://mbed.com/daplink
Unique ID: 99013601066cff495751835187024435a5a5a5a597969908
HIC ID: 97969908
Auto Reset: 1
Automation allowed: 1
Overflow detection: 1
Daplink Mode: Interface
Interface Version: 0254
Bootloader Version: 0254
Git SHA: db711ec68a861b9d9b0d7a7a82071796ec117687
Local Mods: 1
USB Interfaces: MSD, CDC, HID, WebUSB
Bootloader CRC: 0x0697f838
Interface CRC: 0x073fb4fd
Remount count: 0
URL: https://microbit.org/device/?id=9901&v=0254

As you may see it has 9901 in UID which refers to micro:bit board_id. However HIC ID is 97969908 which refers to STM32F103XB.

If I start that micro:bit in MAINTENANCE mode the details.txt is:

# DAPLink Firmware - see https://mbed.com/daplink
Unique ID: 00000000066cff495751835187024435a5a5a5a597969908
HIC ID: 97969908
Auto Reset: 0
Automation allowed: 1
Overflow detection: 0
Daplink Mode: Bootloader
Bootloader Version: 0254
Interface Version: 0254
Git SHA: db711ec68a861b9d9b0d7a7a82071796ec117687
Local Mods: 1
USB Interfaces: MSD
Bootloader CRC: 0x0697f838
Interface CRC: 0x073fb4fd
Remount count: 0
URL: https://mbed.com/daplink

You may see that UID starts from 0000.

Another one counterfeit micro:bit is stuck in MAINTENANCE mode after attempt to flash it with original micro:bit FW because I had no idea that the product is counterfeit. His details.txt is:

# DAPLink Firmware - see https://mbed.com/daplink
Unique ID: 00000000066bff565349655287032725a5a5a5a597969908
HIC ID: 97969908
Auto Reset: 0
Automation allowed: 1
Overflow detection: 0
Daplink Mode: Bootloader
Bootloader Version: 0254
Git SHA: db711ec68a861b9d9b0d7a7a82071796ec117687
Local Mods: 1
USB Interfaces: MSD
Bootloader CRC: 0x0697f838
Interface CRC: 0x4915d882
Remount count: 0
URL: https://mbed.com/daplink

You may see that Interface CRC's are different.

I digged a bit into DAPLink source code to find a solution and have few questions:

Does it mean that both micro:bits have flashed boatloader with stm32f103xb_bl?

When I try to flash kl26z_microbit_if.hex FAIL.TXT shows: 
"error: In application programming aborted due to an out of bounds address.
type: interface" 
I think that caused by the fact that STM32 has DAPLINK_ROM_START 0x08000000, but micro:bit has DAPLINK_ROM_START 0x00000000. Does it mean that I have to change ROM offsets in daplink_addr.h for kl26z to offsets from daplink_addr.h for stm32 and recompile the hex file?

Is it OK to flash 0257 interface to 0254 bootloader?

What can I investigate from Interface CRC?

Regards,

**Comment by andreipavlevich** [comment]
I tried to change IF ROM addresses but it looks like micro:bit IF FW is 95Kb (0x17C00), but with STM32 bootloader there are only 79Kb (0x13C00) available for the IF ROM. So that's mean that I need to reflash bootloader with KL26Z_BL.HEX. However counterfeit micro:bit worked well with STM32 BL. 

Does anybody have an idea what to do with KL26Z_MICROBIT_IF source to make it possible to flash it at 0x0800C000?

**Reply by andreipavlevich** [reply]
I have some progress but still stuck. I managed to modify 0254_kl26z_microbit_0x8000.hex which I downloaded here. What I did:
1. I converted it to bin using hex2bin.py
2. I build a new hex with 0x0800C000 offset using post_build_script
3. I modified in that hex data 01 99 96 97 to 08 99 96 97 (l believe it is HIC ID data) and also modified checksum.
4. I deleted everything that was below :10FBF000

Now it looks like the firmware have flashed, but it still unable to load in Interface mode:

# DAPLink Firmware - see https://mbed.com/daplink
Unique ID: 00000000066bff565349655287032725a5a5a5a597969908
HIC ID: 97969908
Auto Reset: 0
Automation allowed: 1
Overflow detection: 0
Daplink Mode: Bootloader
Bootloader Version: 0254
Interface Version: 0254
Git SHA: db711ec68a861b9d9b0d7a7a82071796ec117687
Local Mods: 1
USB Interfaces: MSD
Bootloader CRC: 0x0697f838
Interface CRC: 0xd97f9a86
Remount count: 0
URL: https://mbed.com/daplink

Any idea please what do to next?

**Comment by andreipavlevich** [comment]
I tried to build IF firmware:

Changes in projects.yaml:

    microbit_clone_if:
        - *module_if
        - *module_hic_stm32f103xb
        - records/board/microbit.yaml

Microbit.yaml:

common:
    macros:
        - HID_LED_DEF=GPIO_LED_ON
        - CDC_LED_DEF=GPIO_LED_ON
        - MSC_LED_DEF=GPIO_LED_ON
        - USB_PROD_STR="BBC micro:bit CMSIS-DAP"
        - CDC_ACM_DEFAULT_BAUDRATE=115200
    sources:
        board:
            - source/board/microbit.c
        family:
            - source/family/nordic/nrf51822/target_nrf51.c
            - source/family/nordic/target_reset_nrf51.c

Changes in microbit.c:

#include "IO_Config.h"
#include "DAP.h"
#include "target_family.h"
#include "target_board.h"
extern target_cfg_t target_device_nrf51822_16;
const board_info_t g_board_info = {
    .board_id = "9901",
    .info_version = kBoardInfoVersion,
    .family_id = kNordic_Nrf51_FamilyID,
    .daplink_url_name = "MICROBITHTM",
    .daplink_drive_name = "MICROBIT   ",
    .daplink_target_url = "https://microbit.org/device/?id=@B&v=@V",
    .target_cfg = &target_device_nrf51822_16,
    .board_vendor = "Micro:bit Educational Foundation",
    .board_name = "BBC micro:bit V1",
};

I compiled and flashed it well and DETAILS.TXT now is:

# DAPLink Firmware - see https://daplink.io
Build ID: 83a882d (gcc, local mods)
Unique ID: 99013601066bff565349655287032725a5a5a5a597969908
HIC ID: 97969908
Auto Reset: 1
Automation allowed: 1
Overflow detection: 1
Incompatible image detection: 0
Page erasing: 0
Daplink Mode: Interface
Interface Version: 0258
Bootloader Version: 0254
Git SHA: 83a882db2e6b16b839668b406b0c5df7b20d84ea
Local Mods: 1
USB Interfaces: MSD, CDC, HID, WebUSB
Bootloader CRC: 0x0697f838
Interface CRC: 0xbba6a2a2
Remount count: 0
URL: https://microbit.org/device/?id=9901&v=0258

However it doesn't work with any HEX file produced by Makecode, FAIL.TXT is:
error: In application programming failed because the update sent was incomplete.
type: interface

What did I wrong? Any idea please?

**Comment by gangbro** [comment, marked as answer]
[fake_Microbit  V1.1 zip](https://github.com/ARMmbed/DAPLink/files/8611525/fake_Microbit.V1.1.zip)
Try this 。zip 2 bin

**Reply by andreipavlevich** [reply]
Thank you for the file! 
It has 0x0000 offset, so I used post_build_script to make 0x0800C000 offset and it worked out!

# DAPLink Firmware - see https://mbed.com/daplink
Unique ID: 99013601066bff565349655287032725a5a5a5a597969908
HIC ID: 97969908
Auto Reset: 1
Automation allowed: 1
Overflow detection: 1
Page erasing: 0
Daplink Mode: Interface
Interface Version: 0254
Bootloader Version: 0254
Git SHA: db711ec68a861b9d9b0d7a7a82071796ec117687
Local Mods: 1
USB Interfaces: MSD, CDC, HID, WebUSB
Bootloader CRC: 0x0697f838
Interface CRC: 0x48754871
Remount count: 2
URL: https://microbit.org/device/?id=9901&v=@V

What actually did you do with a source files? Can you share it please?

**Reply by gangbro** [reply]
This file was given to me by the person who sold me the stuff.
I'm a newbie and can't help much

**Reply by gangbro** [reply]
![image](https://user-images.githubusercontent.com/52055723/166506805-d3deb335-daf0-4d7f-a275-4f76549adf20.png)

**Reply by andreipavlevich** [reply]
bin file you provided doesn't have any signs of kl26z and nrf51822

**Reply by gangbro** [reply]
![image](https://user-images.githubusercontent.com/52055723/166613495-4466e78c-f9ec-4e09-a6b5-67b63994a1a6.png)
I use jlink to set the kl26 model to connect the device, but there is an error in it, so as you said, it should be an STM32, not a KL26

**Reply by gangbro** [reply]
I also did not find the corresponding BGA model product in the official datasheet database. So it might be a remarked fake IC

**Reply by andreipavlevich** [reply]
I have 2 counterfeit micro:bits, both have marked kl26z and nrf51822. I don't know if it's possible to flash kl26z with stm32f102 bootloader for some reason. The device is weird.

**Reply by rodrigopdcouto** [reply]
i have the same problem

**Reply by rodrigopdcouto** [reply]
could you give me the hex file?

**Reply by rodrigopdcouto** [reply]
Thank you very much for your help Andrei, unfortunately after sending the hex file to the microbit the error is now different.

FAIL.TXT --> error: The starting address for the interface update is wrong.
type: user

**Reply by rodrigopdcouto** [reply]
# DAPLink Firmware - see https://mbed.com/daplink
Unique ID: 00000000066aff555251675287155314a5a5a5a597969908
HIC ID: 97969908
Auto Reset: 0
Automation allowed: 1
Overflow detection: 0
Daplink Mode: Bootloader
Bootloader Version: 0254
Git SHA: db711ec68a861b9d9b0d7a7a82071796ec117687
Local Mods: 1
USB Interfaces: MSD
Bootloader CRC: 0x0697f838
Interface CRC: 0x4915d882
Remount count: 0
URL: https://mbed.com/daplink

**Reply by andreipavlevich** [reply]
i reaploded the file. Did you try the latest?

**Reply by rodrigopdcouto** [reply]
I'm newbie and inexperienced in the subject, I bought this microbit and when trying to update I found it to be fake. What would the re-installation procedure look like? if you can point me to some article to understand...

**Reply by andreipavlevich** [reply]
[fake_microbit_v15_0254.zip](https://github.com/ARMmbed/DAPLink/files/8693016/fake_microbit_v15_0254.zip)

Unzip and upload this file to micro:bit drive in MAINTENANCE mode. It would work if you have a micro:bit with stm32 bootloader (HIC ID 97969908)

**Reply by rodrigopdcouto** [reply]
My friend, it's workinggggg, uhullll!!!! you are the best. thx! thx!

**Reply by rodrigopdcouto** [reply]
# DAPLink Firmware - see https://mbed.com/daplink
Unique ID: 99013601066aff555251675287155314a5a5a5a597969908
HIC ID: 97969908
Auto Reset: 1
Automation allowed: 1
Overflow detection: 1
Page erasing: 0
Daplink Mode: Interface
Interface Version: 0254
Bootloader Version: 0254
Git SHA: db711ec68a861b9d9b0d7a7a82071796ec117687
Local Mods: 1
USB Interfaces: MSD, CDC, HID, WebUSB
Bootloader CRC: 0x0697f838
Interface CRC: 0x48754871
Remount count: 0
URL: https://microbit.org/device/?id=9901&v=@V

**Reply by rodrigopdcouto** [reply]
everything is working as it should!!!

**Reply by andreipavlevich** [reply]
it would be great if somebody experienced from this project would try to analyse the bin-file to find out what kind of commits were made in a source code.

**Reply by rodrigopdcouto** [reply]
Agree, an experienced developer would be very helpful in figuring out the code changes. But what I find bad about this whole situation is the fact that I bought a fake microbit and went through this whole complicated situation of seeing the device get damaged when trying to update. Thank you for making the file available to bring my microbit back to life.

**Reply by gangbro** [reply]
The webusb port is available

**Reply by Athauelpa** [reply]
The fake file worked

**Reply by rodrigopdcouto** [reply]
[fake_Microbit_V1.1.zip](https://www.mediafire.com/file/m41o34rf8pk9a42/fake_Microbit_V1.1.zip/file) Try this file.

**Comment by rodrigopdcouto** [comment]
[fake_Microbit_V1.1.zip](https://www.mediafire.com/file/m41o34rf8pk9a42/fake_Microbit_V1.1.zip/file) Try this file.

**Comment by Athauelpa** [comment]
Thanks,
Its working again
Greetings Benny

Op za 21 jan. 2023 16:35 schreef Rodrigo Couto ***@***.***>:

**Comment by voyagerintheshell** [comment]
eyyy thank you man, but just a heads up for people on linux the first zip provided by gangbro isnt unzippable returns the error `Archive:  fake_Microbit.V1.1.zip
  End-of-central-directory signature not found.  Either this file is not
  a zipfile, or it constitutes one disk of a multi-part archive.  In the
  latter case the central directory and zipfile comment will be found on
  the last disk(s) of this archive.
unzip:  cannot find zipfile directory in one of fake_Microbit.V1.1.zip or
        fake_Microbit.V1.1.zip.zip, and cannot find fake_Microbit.V1.1.zip.ZIP, period.`

**Comment by petzolkri-design** [comment]
Nagyon köszönöm. Nekem is ez volt a problémám egy 1.5 micorbittel és már mindent kiprobáltam semmi. Ez a [ake_Microbit_V1.1.zip](https://www.mediafire.com/file/m41o34rf8pk9a42/fake_Microbit_V1.1.zip/file) megoldotta egyből.
